### PR TITLE
nixos/dnsmasq: Use attrs instead of plain text config

### PR DESCRIPTION
**Pull request by KoviRobi**
This should make it easier to configure in multiple places, override defaults, etc. Not sure if change is significant enough to warrant an entry in NixOS release notes.

I guess NixOS/rfcs#42 is related.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/dnsmasq.nix [hunk 1]
```diff
@@ -63,6 +73,25 @@ in
         '';
       };
 
+      config = mkOption {
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/dnsmasq.nix:76]: `settings` is the more used name for this

File: nixos/modules/services/networking/dnsmasq.nix [hunk 2]
```diff
@@ -81,6 +110,14 @@ in
 
   config = mkIf cfg.enable {
 
+    warnings = lib.optional (cfg.extraConfig != "") "Dnsmasq now supports `services.dnsmasq.config` for an attribute-set based config";
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/dnsmasq.nix:113]: we should be more specific maybe and say this is going to be deprecated?

Comment (KoviRobi) [reply nixos/modules/services/networking/dnsmasq.nix]: I've made a couple of changes, including this, which I've also now mentioned in the release note (if we are going to deprecate something  thought it should be mentioned).

**Review by RaitoBezarius — APPROVED** [review]
LGTM

**Comment by RaitoBezarius** [comment]
Could you move the release notes to 23.05 now?

**Comment by KoviRobi** [comment]
Of course!

Idle thought, I suspect there's going to be plenty of conflicts like this, until the release note grows big enough that the insert in arbitrary position works well for the merge resolution. I wonder if we could have the release note options included as files, as a glob (e.g. `release-notes/23.05/*.md`), adding new files wouldn't have conflicts like this?

**Review by RaitoBezarius — APPROVED** [review]

**Comment by KoviRobi** [comment]
Whoops, I realised I forgot to bump the version in the release notes when I moved it over, done now.

**Comment by RaitoBezarius** [comment]
@KoviRobi Alas more conflicts. :'( ; please resolve them and I will merge so we can finally enjoy RFC42 on this :D.

**Comment by KoviRobi** [comment]
No worries, and sorry it's taking so long

**Comment by RaitoBezarius** [comment]
> No worries, and sorry it's taking so long

I am the one sorry :)

**Comment by RaitoBezarius** [comment]
@ofborg test schleuder dnscrypt-proxy2

**Comment by KoviRobi** [comment]
@RaitoBezarius No worries, I know everyone has been very busy with 22.11, plus also it's winter so I don't know about you but I get less done due to less sunlight. (But also sadly I have been struggling to be as active for Nix stuff in the past two and a half years)

**Comment by KoviRobi** [comment]
Looks like that didn't work? Let's try that again

@ofborg test schleuder dnscrypt-proxy2

**Comment by RaitoBezarius** [comment]
Thank you a lot @KoviRobi ! (indeed, I miss spring!)

**Comment by c4710n** [comment]
@KoviRobi Sorry to bother you. But how to express the following config after removing `extraConfig`?
```
interface=lo
interface=eth0

address=/example-site.private/10.128.0.1
address=/example-service.private/10.128.0.2
```

**Comment by c4710n** [comment]
> attribute set of (atom (null, bool, int, float or string) or **a list of them for duplicate keys)**

Ah, found it. Sorry to bother you, again. ;)

**Pull request merged**